src/frequenz/sdk/timeseries/_ringbuffer.py

```diff
@@ -0,0 +1,430 @@
+
+        Returns:
+            Index where the value for the given timestamp can be found.
+        """
+        timestamp = self._normalize_timestamp(timestamp)
+
+        if not allow_outside_range and (
+            self._datetime_newest + self._sampling_period < timestamp
+            or timestamp < self._datetime_oldest
+        ):
+            raise IndexError(
+                f"Requested timestamp {timestamp} is "
+                f"outside the range [{self._datetime_oldest} - {self._datetime_newest}]"
+            )
+
+        return self._wrap(
+            round(
+                (timestamp - self._time_index_alignment).total_seconds()
+                / self._sampling_period.total_seconds()
+            )
+        )
+
+    def window(
+        self, start: datetime, end: datetime, force_copy: bool = False
+    ) -> Container:
+        """Request a view on the data between start timestamp and end timestamp.
+
+        If the data is not used immediately it could be overwritten.
+        Always request a copy if you keep the data around for longer.
+
+        Will return a copy in the following cases:
+        * The requested time period is crossing the start/end of the buffer.
+        * The requested time period contains missing entries.
+        * The force_copy parameter was set to True (default False).
+
+        The first case can be avoided by using the appropriate
+        time_index_alignment value in the constructor so that the data lines up
+        with the start/end of the buffer.
+
+        This means, if the caller needs to modify the data to account for
+        missing entries, they can safely do so.
+
+        Args:
+            start: start time of the window.
+            end: end time of the window.
+            force_copy: optional, default False. If True, will always create a
+                copy of the data.
+
+        Raises:
+            IndexError: when requesting a window with invalid timestamps.
+
+        Returns:
+            The requested window
+        """
+        if start > end:
+            raise IndexError(
+                f"end parameter {end} has to predate start parameter {start}"
+            )
+
+        start_index = self.datetime_to_index(start)
+        end_index = self.datetime_to_index(end)
+
+        # Requested window wraps around the ends
+        if start_index >= end_index:
+            window: Any = self._buffer[start_index:]
+
+            if end_index > 0:
+                if isinstance(self._buffer, list):
+                    window += self._buffer[0:end_index]
+                else:
+                    window = np.concatenate((window, self._buffer[0:end_index]))
+            return window
+
+        def in_gap(gap) -> bool:
+            return gap.contains(start) or gap.contains(end)
+
+        # Return a copy if there are none-values in the data
+        if force_copy or any(map(in_gap, self._gaps)):
+            return deepcopy(self._buffer[start_index:end_index])
```

Comment:
I'm not sure `deepcopy()` is necessary nor desired. You can do a shallow copy by using `copy()` instead. Why was a deep copy preferred?